# Pandas

Explore the data set and study the statistical parameters of the columns. Find the columns with the highest absolute correlation with the target column and explain why you think so happens.

In [383]:
import pandas as pd

df = pd.read_csv('data/wine-quality.csv')

In [384]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   type                  6497 non-null   object 
 1   fixed acidity         6497 non-null   float64
 2   volatile acidity      6497 non-null   float64
 3   citric acid           6497 non-null   float64
 4   residual sugar        6497 non-null   float64
 5   chlorides             6497 non-null   float64
 6   free sulfur dioxide   6497 non-null   float64
 7   total sulfur dioxide  6497 non-null   float64
 8   density               6497 non-null   float64
 9   pH                    6497 non-null   float64
 10  sulphates             6497 non-null   float64
 11  alcohol               6497 non-null   float64
 12  quality               6497 non-null   int64  
dtypes: float64(11), int64(1), object(1)
memory usage: 660.0+ KB


In [385]:
df.head()

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0  white            7.0              0.27         0.36            20.7   
1  white            6.3              0.30         0.34             1.6   
2  white            8.1              0.28         0.40             6.9   
3  white            7.2              0.23         0.32             8.5   
4  white            7.2              0.23         0.32             8.5   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.045                 45.0                 170.0   1.0010  3.00   
1      0.049                 14.0                 132.0   0.9940  3.30   
2      0.050                 30.0                  97.0   0.9951  3.26   
3      0.058                 47.0                 186.0   0.9956  3.19   
4      0.058                 47.0                 186.0   0.9956  3.19   

   sulphates  alcohol  quality  
0       0.45      8.8        6  
1       0.49      9.5        6  
2       0.44     10.1        6  
3       0.40      9.9        6  
4       0.40      9.9        6

In [386]:
df['type'].value_counts()

white    4898
red      1599
Name: type, dtype: int64

In [387]:
# Change the type of the type column to category
# 1 - white, 0 - red
df['type'] = df['type'].astype('category').cat.codes

df.head()

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0     1            7.0              0.27         0.36            20.7   
1     1            6.3              0.30         0.34             1.6   
2     1            8.1              0.28         0.40             6.9   
3     1            7.2              0.23         0.32             8.5   
4     1            7.2              0.23         0.32             8.5   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.045                 45.0                 170.0   1.0010  3.00   
1      0.049                 14.0                 132.0   0.9940  3.30   
2      0.050                 30.0                  97.0   0.9951  3.26   
3      0.058                 47.0                 186.0   0.9956  3.19   
4      0.058                 47.0                 186.0   0.9956  3.19   

   sulphates  alcohol  quality  
0       0.45      8.8        6  
1       0.49      9.5        6  
2       0.44     10.1        6  
3       0.40      9.9        6  
4       0.40      9.9        6

In [388]:
df.describe()

type  fixed acidity  volatile acidity  citric acid  \
count  6497.000000    6497.000000       6497.000000  6497.000000   
mean      0.753886       7.215307          0.339666     0.318633   
std       0.430779       1.296434          0.164636     0.145318   
min       0.000000       3.800000          0.080000     0.000000   
25%       1.000000       6.400000          0.230000     0.250000   
50%       1.000000       7.000000          0.290000     0.310000   
75%       1.000000       7.700000          0.400000     0.390000   
max       1.000000      15.900000          1.580000     1.660000   

       residual sugar    chlorides  free sulfur dioxide  total sulfur dioxide  \
count     6497.000000  6497.000000          6497.000000           6497.000000   
mean         5.443235     0.056034            30.525319            115.744574   
std          4.757804     0.035034            17.749400             56.521855   
min          0.600000     0.009000             1.000000              6.000000   
25%          1.800000     0.038000            17.000000             77.000000   
50%          3.000000     0.047000            29.000000            118.000000   
75%          8.100000     0.065000            41.000000            156.000000   
max         65.800000     0.611000           289.000000            440.000000   

           density           pH    sulphates      alcohol      quality  
count  6497.000000  6497.000000  6497.000000  6497.000000  6497.000000  
mean      0.994697     3.218501     0.531268    10.491801     5.818378  
std       0.002999     0.160787     0.148806     1.192712     0.873255  
min       0.987110     2.720000     0.220000     8.000000     3.000000  
25%       0.992340     3.110000     0.430000     9.500000     5.000000  
50%       0.994890     3.210000     0.510000    10.300000     6.000000  
75%       0.996990     3.320000     0.600000    11.300000     6.000000  
max       1.038980     4.010000     2.000000    14.900000     9.000000

In [389]:
df.median()

type                      1.00000
fixed acidity             7.00000
volatile acidity          0.29000
citric acid               0.31000
residual sugar            3.00000
chlorides                 0.04700
free sulfur dioxide      29.00000
total sulfur dioxide    118.00000
density                   0.99489
pH                        3.21000
sulphates                 0.51000
alcohol                  10.30000
quality                   6.00000
dtype: float64

In [390]:
df.var()

type                       0.185570
fixed acidity              1.680740
volatile acidity           0.027105
citric acid                0.021117
residual sugar            22.636696
chlorides                  0.001227
free sulfur dioxide      315.041192
total sulfur dioxide    3194.720039
density                    0.000009
pH                         0.025853
sulphates                  0.022143
alcohol                    1.422561
quality                    0.762575
dtype: float64

In [392]:
df.groupby('type').mean()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
type                                                                            
0          8.319637          0.527821     0.270976        2.538806   0.087467   
1          6.854788          0.278241     0.334192        6.391415   0.045772   

      free sulfur dioxide  total sulfur dioxide   density        pH  \
type                                                                  
0               15.874922             46.467792  0.996747  3.311113   
1               35.308085            138.360657  0.994027  3.188267   

      sulphates    alcohol   quality  
type                                  
0      0.658149  10.422983  5.636023  
1      0.489847  10.514267  5.877909

In [393]:
df.groupby('type').median()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
type                                                                            
0               7.9              0.52         0.26             2.2      0.079   
1               6.8              0.26         0.32             5.2      0.043   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
type                                                                        
0                    14.0                  38.0  0.99675  3.31       0.62   
1                    34.0                 134.0  0.99374  3.18       0.47   

      alcohol  quality  
type                    
0        10.2      6.0  
1        10.4      6.0

In [394]:
# Mode of each column for each wine type
df.groupby('type').agg(lambda x: x.value_counts().index[0])

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
type                                                                            
0               7.2              0.60          0.0             2.0      0.080   
1               6.8              0.28          0.3             1.2      0.044   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
type                                                                        
0                     6.0                  28.0   0.9972  3.30        0.6   
1                    29.0                 111.0   0.9920  3.14        0.5   

      alcohol  quality  
type                    
0         9.5        5  
1         9.4        6

In [395]:
df.groupby('type').agg(['min', 'max'])

fixed acidity       volatile acidity       citric acid        \
               min   max              min   max         min   max   
type                                                                
0              4.6  15.9             0.12  1.58         0.0  1.00   
1              3.8  14.2             0.08  1.10         0.0  1.66   

     residual sugar       chlorides         ...  density             pH        \
                min   max       min    max  ...      min      max   min   max   
type                                        ...                                 
0               0.9  15.5     0.012  0.611  ...  0.99007  1.00369  2.74  4.01   
1               0.6  65.8     0.009  0.346  ...  0.98711  1.03898  2.72  3.82   

     sulphates       alcohol       quality      
           min   max     min   max     min max  
type                                            
0         0.33  2.00     8.4  14.9       3   8  
1         0.22  1.08     8.0  14.2       3   9  

[2 rows x 24 columns]

In [396]:
df.groupby('type').std()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
type                                                                            
0          1.741096          0.179060     0.194801        1.409928   0.047065   
1          0.843868          0.100795     0.121020        5.072058   0.021848   

      free sulfur dioxide  total sulfur dioxide   density        pH  \
type                                                                  
0               10.460157             32.895324  0.001887  0.154386   
1               17.007137             42.498065  0.002991  0.151001   

      sulphates   alcohol   quality  
type                                 
0      0.169507  1.065668  0.807569  
1      0.114126  1.230621  0.885639

In [397]:
df.groupby('type').var()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
type                                                                            
0          3.031416          0.032062     0.037947        1.987897   0.002215   
1          0.712114          0.010160     0.014646       25.725770   0.000477   

      free sulfur dioxide  total sulfur dioxide   density        pH  \
type                                                                  
0              109.414884           1082.102373  0.000004  0.023835   
1              289.242720           1806.085491  0.000009  0.022801   

      sulphates   alcohol   quality  
type                                 
0      0.028733  1.135647  0.652168  
1      0.013025  1.514427  0.784356

In [398]:
df.corr()

type  fixed acidity  volatile acidity  citric acid  \
type                  1.000000      -0.486740         -0.653036     0.187397   
fixed acidity        -0.486740       1.000000          0.219008     0.324436   
volatile acidity     -0.653036       0.219008          1.000000    -0.377981   
citric acid           0.187397       0.324436         -0.377981     1.000000   
residual sugar        0.348821      -0.111981         -0.196011     0.142451   
chlorides            -0.512678       0.298195          0.377124     0.038998   
free sulfur dioxide   0.471644      -0.282735         -0.352557     0.133126   
total sulfur dioxide  0.700357      -0.329054         -0.414476     0.195242   
density              -0.390645       0.458910          0.271296     0.096154   
pH                   -0.329129      -0.252700          0.261454    -0.329808   
sulphates            -0.487218       0.299568          0.225984     0.056197   
alcohol               0.032970      -0.095452         -0.037640    -0.010493   
quality               0.119323      -0.076743         -0.265699     0.085532   

                      residual sugar  chlorides  free sulfur dioxide  \
type                        0.348821  -0.512678             0.471644   
fixed acidity              -0.111981   0.298195            -0.282735   
volatile acidity           -0.196011   0.377124            -0.352557   
citric acid                 0.142451   0.038998             0.133126   
residual sugar              1.000000  -0.128940             0.402871   
chlorides                  -0.128940   1.000000            -0.195045   
free sulfur dioxide         0.402871  -0.195045             1.000000   
total sulfur dioxide        0.495482  -0.279630             0.720934   
density                     0.552517   0.362615             0.025717   
pH                         -0.267320   0.044708            -0.145854   
sulphates                  -0.185927   0.395593            -0.188457   
alcohol                    -0.359415  -0.256916            -0.179838   
quality                    -0.036980  -0.200666             0.055463   

                      total sulfur dioxide   density        pH  sulphates  \
type                              0.700357 -0.390645 -0.329129  -0.487218   
fixed acidity                    -0.329054  0.458910 -0.252700   0.299568   
volatile acidity                 -0.414476  0.271296  0.261454   0.225984   
citric acid                       0.195242  0.096154 -0.329808   0.056197   
residual sugar                    0.495482  0.552517 -0.267320  -0.185927   
chlorides                        -0.279630  0.362615  0.044708   0.395593   
free sulfur dioxide               0.720934  0.025717 -0.145854  -0.188457   
total sulfur dioxide              1.000000  0.032395 -0.238413  -0.275727   
density                           0.032395  1.000000  0.011686   0.259478   
pH                               -0.238413  0.011686  1.000000   0.192123   
sulphates                        -0.275727  0.259478  0.192123   1.000000   
alcohol                          -0.265740 -0.686745  0.121248  -0.003029   
quality                          -0.041385 -0.305858  0.019506   0.038485   

                       alcohol   quality  
type                  0.032970  0.119323  
fixed acidity        -0.095452 -0.076743  
volatile acidity     -0.037640 -0.265699  
citric acid          -0.010493  0.085532  
residual sugar       -0.359415 -0.036980  
chlorides            -0.256916 -0.200666  
free sulfur dioxide  -0.179838  0.055463  
total sulfur dioxide -0.265740 -0.041385  
density              -0.686745 -0.305858  
pH                    0.121248  0.019506  
sulphates            -0.003029  0.038485  
alcohol               1.000000  0.444319  
quality               0.444319  1.000000

In [399]:
target = 'quality'

df.groupby(target).mean()

type  fixed acidity  volatile acidity  citric acid  \
quality                                                           
3        0.666667       7.853333          0.517000     0.281000   
4        0.754630       7.288889          0.457963     0.272315   
5        0.681478       7.326801          0.389614     0.307722   
6        0.775035       7.177257          0.313863     0.323583   
7        0.815570       7.128962          0.288800     0.334764   
8        0.906736       6.835233          0.291010     0.332539   
9        1.000000       7.420000          0.298000     0.386000   

         residual sugar  chlorides  free sulfur dioxide  total sulfur dioxide  \
quality                                                                         
3              5.140000   0.077033            39.216667            122.033333   
4              4.153704   0.060056            20.636574            103.432870   
5              5.804116   0.064666            30.237371            120.839102   
6              5.549753   0.054157            31.165021            115.410790   
7              4.731696   0.045272            30.422150            108.498610   
8              5.382902   0.041124            34.533679            117.518135   
9              4.120000   0.027400            33.400000            116.000000   

          density        pH  sulphates    alcohol  
quality                                            
3        0.995744  3.257667   0.506333  10.215000  
4        0.994833  3.231620   0.505648  10.180093  
5        0.995849  3.212189   0.526403   9.837783  
6        0.994558  3.217726   0.532549  10.587553  
7        0.993126  3.228072   0.547025  11.386006  
8        0.992514  3.223212   0.512487  11.678756  
9        0.991460  3.308000   0.466000  12.180000

In [400]:
# Find what columns have the highest absolute correlation with the target column
correlations = df.corr()[target].abs().sort_values(ascending=False)
correlations.drop(target, inplace=True)

print(f'Correlations: \n{correlations}\n')

Correlations: 
alcohol                 0.444319
density                 0.305858
volatile acidity        0.265699
chlorides               0.200666
type                    0.119323
citric acid             0.085532
fixed acidity           0.076743
free sulfur dioxide     0.055463
total sulfur dioxide    0.041385
sulphates               0.038485
residual sugar          0.036980
pH                      0.019506
Name: quality, dtype: float64



## Conclusion

Based on the correlation matrix and other statistical parameters, the top 3 numerical columns with the highest absolute correlation with the target column are: 

1. Alcohol

The correlation between the alcohol and the quality of the wine is obvious. The higher the alcohol content, the better the quality of the wine.

2. Density

There is a negative correlation between density and wine quality. High-density wines often have more residual sugar and less alcohol, which can result in a less desirable wine. So, the correct conclusion would be: the lower the density, the better the quality of the wine.

3. Volatile acidity

In terms of the volatile acidity, the higher the volatile acidity, the lower the quality of the wine. The volatile acidity is the amount of acetic acid in wine, which at too high of levels can lead to an unpleasant, vinegar taste.

The correlation between wine type and quality is relatively low, with a correlation coefficient of 0.119323. This suggests that wine type has a weak association with wine quality.
In this case, the type of wine (red or white) is not a strong predictor of wine quality. Other physicochemical properties, such as alcohol content, density, and volatile acidity, have a stronger influence on wine quality.